include/range/v3/experimental/utility/generator.hpp

```diff
@@ -73,41 +83,71 @@ namespace ranges
     {
         // An owning coroutine_handle
         template<typename Promise>
-        struct coroutine_owner
-          : std::experimental::coroutine_handle<Promise>
+        struct RANGES_EMPTY_BASES coroutine_owner
+          : private std::experimental::coroutine_handle<Promise>
           , private detail::coroutine_owner_::adl_hook
         {
+            CPP_assert(DerivedFrom<Promise, enable_coroutine_owner>);
             using base_t = std::experimental::coroutine_handle<Promise>;
 
             using base_t::operator bool;
             using base_t::done;
             using base_t::promise;
 
-            constexpr coroutine_owner() noexcept = default;
+            coroutine_owner() = default;
             explicit constexpr coroutine_owner(base_t coro) noexcept
               : base_t(coro)
             {}
             constexpr /*c++14*/ coroutine_owner(coroutine_owner &&that) noexcept
               : base_t(ranges::exchange(that.base(), {}))
+              , copied_(that.copied_.load(std::memory_order_relaxed))
             {}
-
-            ~coroutine_owner() { if (*this) base().destroy(); }
-
-            coroutine_owner &operator=(coroutine_owner &&that) noexcept
+            constexpr /*c++14*/ coroutine_owner(coroutine_owner const &that) noexcept
+              : base_t(that.handle())
+              , copied_(that.handle() != nullptr)
+            {
+                if(*this)
+                {
+                    that.copied_.store(true, std::memory_order_relaxed);
+                    base().promise().refcount_.fetch_add(1, std::memory_order_relaxed);
+                }
+            }
+            ~coroutine_owner()
+            {
+                if(base() && (!copied_.load(std::memory_order_relaxed) ||
```

Comment:
`relaxed` is too weak here to guarantee that relaxed stores from other threads happen before this load, no?

Reply:
@lewissbaker When you are back from vacation, can you chime in here since you talked me into making these `relaxed`?